dev/src/cubegen.cpp

```diff
@@ -589,15 +601,102 @@ nfr("cg_load_vox", "name", "S", "R:voxels]S?",
+                    auto c = (const uint8_t*)p;
+                    auto node_id = ReadMemInc<int32_t>(c);
+                    auto dict_len = ReadMemInc<int32_t>(c);
+                    for (int i = 0; i<dict_len; ++i) {
+                        string key, value;
+                        ReadVec32(c, key);
+                        ReadVec32(c, value);
+                        if (key == "_name")
+                            node_names.insert_or_assign(node_id, value);
+                    }
+                    auto child_num = ReadMemInc<int32_t>(c);
+                    for (int i = 0; i<child_num; ++i) {
+                        auto child_node_id = ReadMemInc<int32_t>(c);
+                        node_graph.insert_or_assign(child_node_id, node_id);
+                    }
+                } else if (!strncmp(id, "nSHP", 4)) {
+                    auto c = (const uint8_t*)p;
+                    auto node_id = ReadMemInc<int32_t>(c);
+                    auto dict_len = ReadMemInc<int32_t>(c);
+                    for (int i = 0; i<dict_len; ++i) {
+                        string key, value;
+                        ReadVec32(c, key);
+                        ReadVec32(c, value);
+                        if (key == "_name")
+                            node_names.insert_or_assign(node_id, value);
+                    }
+                    auto models_num = ReadMemInc<int32_t>(c);
+                    for (int i = 0; i<models_num; ++i) {
+                        auto model_id = ReadMemInc<int32_t>(c);
+                        node_to_model.insert_or_assign(node_id, model_id);
+                        auto dict_len = ReadMemInc<int32_t>(c);
+                        for (int i = 0; i<dict_len; ++i) {
+                            string key, value;
+                            ReadVec32(c, key);
+                            ReadVec32(c, value);
+                        }
+                        auto reserved = ReadMemInc<int32_t>(c);
+                    }
+                } else if (!strncmp(id, "LAYR", 4)) {
+                    // Layer metadata
+                    auto c = (const uint8_t*)p;
+                    auto layer_id = ReadMemInc<int32_t>(c);
+                    auto dict_len = ReadMemInc<int32_t>(c);
+                    for (int i = 0; i<dict_len; ++i) {
+                        string key, value;
+                        ReadVec32(c, key);
+                        ReadVec32(c, value);
+                        if (key == "_name")
+                            layer_names.insert_or_assign(layer_id, value);
+                    }
                 } else {
                     chunks_skipped = true;
                 }
                 p += contentlen;
             }
+            for (auto &i : node_to_layer)
+                if (layer_names.contains(i.second) && !node_names.contains(i.first))
+                    node_names.insert_or_assign(i.first, layer_names[i.second]);
+            for (auto &i : node_to_model) {
```

Comment:
So all those maps are needed to just generate a single string? I mean, it's nice to have this code in the sense that we may later want to do more with this hierarchy data, but I would not want it to slow down `.vox` loading.. a `map` can imply a lot of dynamic allocation, any way you can represent this more efficiently?

Also, how does one set a name in the MV UI? Couldn't figure that last time I checked.

Reply:
Annoyingly the data is hidden behind model_ids, layer_ids and node_ids who all need to be connected together and all of which can have string->string metadata sections.

Perhaps i should put this resolving behind a setting/bool argument for code that requests this data ?

It is a feature in the latest beta 0.99.7.0

Reply:
yeah a bool could work, or alternatively only parse it if the file contains multiple models, that would also reduce the amount of files for which it needs to happen.